src/main/scala/org/apache/spark/sql/pulsar/PulsarSource.scala

```diff
@@ -54,6 +68,8 @@ private[pulsar] class PulsarSource(
 
   private var currentTopicOffsets: Option[Map[String, MessageId]] = None
 
+  private lazy val pulsarAdmin = PulsarAdmin.builder().serviceHttpUrl(serviceUrl).build()
```

Comment:
Can you move this object along with the admission control logic into pulsar helper?

Reply:
Done